File: Day4/day4.ts
import { readFileSync } from 'fs';

class Square {
    value: number;
    marked: boolean = false;

    constructor(value) {
        this.value = value;
    }
}

type Board = {
    squares: Square[][];
}

export function partOne(input: string[]): number {
    let { numbers, boards, squareMap } = parseInput(input);

    // console.log(squareMap);

    for (let i = 0; i < numbers.length; i++) {
        let num = numbers[i];
    // numbers.forEach((num, i) => {
        // console.log(num);
        // console.log(squareMap.get(`${num}`));
        let squares = squareMap.get(`${num}`);
        for (let j = 0; j < squares.length; j++) {
            squares[j].marked = true;
        // squares.forEach(sq => {
            // sq.marked = true;
        // });
        }
        squareMap.set(`${num}`, squares);

        if (i > 5) {
            let score = checkWin(boards, num);
            if (score != -1) {
                console.log(score);
                return score;
            }
        }
    // });
    }
    
    return -1;
}

function parseInput(input: string[]): { numbers: number[], boards: Board[], squareMap: Map<string, Square[]> } {
    let numbers = input[0].split(',').map(val => parseInt(val));
    let boards: Board[] = [];
    let squareMap = new Map<string, Square[]>();

    for (let i = 2; i < input.length; i += 6) {
        let squares: Square[][] = [];
        for (let j = i; j < i + 5; j++) {
            let row = input[j].trim().replace(/ +/g, '_').split('_').map(val => new Square(val));
            row.forEach(square => {
                if (squareMap.has(`${square.value}`)) {
                    squareMap.set(`${square.value}`, [...squareMap.get(`${square.value}`), square]);
                } else {
                    squareMap.set(`${square.value}`, [square]);
                }
            });

            squares.push(row);
        }

        boards.push({
            squares: squares
        });
    }

    // console.log('boards count: ', boards.length);

    return { numbers, boards, squareMap };
}

function checkWin(boards: Board[], lastCalled: number): number {
    console.log(boards);
    boards.forEach(board => {
        // check rows
        for (let i = 0; i < board.squares.length; i++) {
            let win = true
            for (let j = 0; j < board.squares[i].length; j++) {
                if (!board.squares[i][j].value) {
                    win = false;
                }
            }

            if (win) {
                // console.log('found winning board: ', board);
                return getScore(board, lastCalled);
            }
        }

        // check columns
        for (let i = 0; i < board.squares.length; i++) {
            let win = true
            for (let j = 0; j < board.squares[i].length; j++) {
                if (!board.squares[j][i].value) {
                    win = false;
                }
            }

            if (win) {
                // console.log('found winning board: ', board);
                return getScore(board, lastCalled);
            }
        }
    });

    return -1;
}

function getScore(board: Board, lastCalled: number): number {
    // let score = board.squares
    //     .flatMap(sq => sq)
    //     .filter(sq => !sq.marked)
    //     .map(sq => sq.value)
    //     .reduce((sum, val) => sum += val, 0) * lastCalled;
    let unmarkedSquares = board.squares.flatMap(sq => sq).filter(sq => !sq.marked);
    console.log('unmarkedsquares: ', unmarkedSquares);
    let sum = unmarkedSquares.map(sq => sq.value).reduce((sum, val) => sum += val, 0);
    console.log('sum: ', sum);
    console.log('lastCalled: ', lastCalled);
    let score = sum * lastCalled;

    console.log(score);
    return score;
}

export function solve() {
    let input = readFileSync('./day4_sample_input.txt', 'utf-8').split('\r\n');

    console.log(partOne(input));
    // console.log(partTwo(input));
}

solve();

File: Day4/day4.dart
import 'dart:io';

class BingoSquare {
  bool marked = false;
  int value;

  BingoSquare(this.value);
}

class Board {
  List<List<BingoSquare>> squares;

  Board(this.squares);
}

class Game {
  List<int> numbers;
  List<Board> boards;
  Map<int, List<BingoSquare>> map;

  Game(this.numbers, this.boards, this.map);
}

void main() async {
  String contents = await File("./day4_sample_input.txt").readAsString();
  List<String> input = contents.split("\n");

  print("Part one: ${partOne(input)}");
  print("Part two: ${partTwo(input)}");
}

int partOne(List<String> input) {
  var game = parseInput(input);

  for (var i = 0; i < game.numbers.length; i++) {
    var number = game.numbers[i];
    var squares = game.map[number];
    if (squares != null) {
      for (var i = 0; i < squares.length; i++) {
        squares[i].marked = true;
      }
    }

    if (i > 5) {
      var score = checkWin(game.boards, number);

      if (score != -1) {
        return score;
      }
    }
  }

  return -1;
}

int partTwo(List<String> input) {
  var game = parseInput(input);

  for (var i = 0; i < game.numbers.length; i++) {
    var number = game.numbers[i];
    var squares = game.map[number];
    if (squares != null) {
      for (var i = 0; i < squares.length; i++) {
        squares[i].marked = true;
      }
    }

    if (i > 5) {
      var score = checkWinPartTwo(game.boards, number);

      if (score != -1) {
        return score;
      }
    }
  }

  return -1;
}

Game parseInput(List<String> input) {
  List<int> numbers = [];
  for (String num in input[0].split(",")) {
    var value = int.parse(num);
    numbers.add(value);
  }

  List<Board> boards = [];
  Map<int, List<BingoSquare>> map = {};

  for (var i = 2; i < input.length; i += 6) {
    List<List<BingoSquare>> squares = [];
    for (var j = i; j < i + 5; j++) {
      List<BingoSquare> squareRow = [];
      var row = input[j].split(RegExp(r" +"));
      for (var numStr in row) {
        var num = int.tryParse(numStr);
        if (num != null) {
          var square = BingoSquare(num);
          if (map.containsKey(num)) {
            var list = map[num]!;
            list.add(square);
            map[num] = list;
          } else {
            map[num] = [square];
          }

          squareRow.add(square);
        }
      }

      squares.add(squareRow);
    }

    print(squares);

    boards.add(new Board(squares));
  }

  return Game(numbers, boards, map);
}

int checkWin(List<Board> boards, int lastCalled) {
  for (var board in boards) {
    for (var i = 0; i < board.squares.length; i++) {
      var horizontalWin = true;
      var verticalWin = true;
      for (var j = 0; j < board.squares[i].length; j++) {
        if (!board.squares[i][j].marked) {
          horizontalWin = false;
        }

        if (!board.squares[j][i].marked) {
          verticalWin = false;
        }
      }

      if (verticalWin || horizontalWin) {
        return getScore(board, lastCalled);
      }
    }
  }

  return -1;
}

int checkWinPartTwo(List<Board> boards, int lastCalled) {
  for (var b = 0; b < boards.length; b++) {
    var board = boards[b];
    for (var i = 0; i < board.squares.length; i++) {
      var horizontalWin = true;
      var verticalWin = true;
      for (var j = 0; j < board.squares[i].length; j++) {
        if (!board.squares[i][j].marked) {
          horizontalWin = false;
        }

        if (!board.squares[j][i].marked) {
          verticalWin = false;
        }
      }

      if (verticalWin || horizontalWin && board.squares.length > 1) {
        boards.removeAt(b);
        b--;
        break;
        // return getScore(board, lastCalled);
      } else if (verticalWin || horizontalWin && board.squares.length == 1) {
        return getScore(board, lastCalled);
      }
    }
  }

  return -1;
}

int getScore(Board board, int lastCalled) {
  var sum = 0;
  for (var i = 0; i < board.squares.length; i++) {
    for (var j = 0; j < board.squares[i].length; j++) {
      if (!board.squares[i][j].marked) {
        sum += board.squares[i][j].value;
      }
    }
  }

  return sum * lastCalled;
}
